bionic/core/flow_execution.py

```diff
@@ -401,31 +408,48 @@ def _mark_entry_active(self, entry):
     def _wait_on_in_progress_entries(self):
         "Waits on any in-progress entry to finish."
         futures = set(entry.future for entry in self._in_progress_entries)
-        if self._parallel_execution_enabled:
-            finished_futures, _ = wait(futures, return_when=FIRST_COMPLETED)
-        else:
-            assert self._aip_execution_enabled
-            # TODO: This works for PoC, but maybe there is a better way
-            # to do this. Also consider making the sleep time configurable.
-            # See this comment for example
-            # https://github.com/square/bionic/pull/253#discussion_r487066911
-            while all(future.running() for future in futures):
-                time.sleep(10)
-            finished_futures = [future for future in futures if future.done()]
-        for finished_future in finished_futures:
-            # TODO: If there is an error, consider waiting till all
-            # futures are done. With AIP execution, we can have one
-            # task fail while others are running. If we don't wait till
-            # those tasks are done, the user might perform another
-            # `get` operation that can spawn the same Task again. Apart
-            # from wasted resource, the other problem is that the older
-            # and newer tasks can potentially conflict writing to the
-            # same cache file.
-            task_keys = finished_future.result()
-            for task_key in task_keys:
-                entry = self._entries_by_task_key[task_key]
-                entry.state.sync_after_remote_computation()
-                assert self._mark_entry_completed_if_possible(entry)
+        finished_futures, _ = wait(futures, return_when=FIRST_COMPLETED)
+        try:
+            for finished_future in finished_futures:
+                task_keys = finished_future.result()
+                for task_key in task_keys:
+                    entry = self._entries_by_task_key[task_key]
+                    entry.state.sync_after_remote_computation()
+                    assert self._mark_entry_completed_if_possible(entry)
+        except Exception as exception:
+            # If there is an error, wait until all futures are done. With AIP
+            # execution, we can have one task fail while others are running. If
+            # we don't wait till those tasks are done, the user might perform
+            # another `get` operation that can spawn the same Task again.
+
+            # Using logger.error instead of logger.exception so that the
+            # exception information can be captured.
+            logger.error(
+                oneline(
+                    f"""
+                    Encountered an error while doing remote computation:
+                    {exception!r}. Waiting for all other remote computation to
+                    complete...
+                    """
+                )
+            )
+            finished_futures, _ = wait(futures, return_when=ALL_COMPLETED)
```

Comment:
There could be futures that failed in the first batch which we haven't had a chance to report, right?

Reply:
Yes. The `wait(...)` here should return the results of all futures, including the one that caused the exception. Hence, the next loop should be able to capture all the exceptions.

Reply:
Ah, right -- got it!